src/hb-ot-cmap-table.hh

```diff
@@ -1348,6 +1352,181 @@ struct cmap
 
+  template<typename Iterable1, typename Iterable2,
+           hb_requires (hb_is_iterator (Iterable1)),
+           hb_requires (hb_is_iterator (Iterable2))>
+  bool compare_iterators (Iterable1 Iter_1, Iterable2 Iter_2) const
+  {
+    if (Iter_1.len () != Iter_2.len ())
+      return false;
+
+    bool is_equal = true;
+    for (auto pair : + hb_zip (Iter_1, Iter_2))
+    {
+      if (pair.first != pair.second)
+      {
+        is_equal = false;
+        break;
+      }
+    }
+    return is_equal;
+  }
+
+  template<typename Iterator,
+	   hb_requires (hb_is_iterator (Iterator))>
+  bool categorize_sub_tables (Iterator encodingrec_iter,
+                              hb_hashmap_t<unsigned, const EncodingRecord *, (unsigned)-1, nullptr>& unicode_bmp_tables, /* INOUT */
+                              hb_hashmap_t<unsigned, const EncodingRecord *, (unsigned)-1, nullptr>& ms_bmp_tables, /* INOUT */
+                              hb_map_t& format12_table_languages /* INOUT */) const
+  {
+    const EncodingRecord *unicode_bmp= nullptr, *unicode_ucs4 = nullptr, *ms_bmp = nullptr, *ms_ucs4 = nullptr;
+    bool has_format12 = false;
+    for (auto _ : encodingrec_iter)
+    {
+      const EncodingRecord& record = _.first;
+      unsigned index = _.second;
+      unsigned format = (this + record.subtable).u.format;
+      const EncodingRecord *table = hb_addressof (record);
+      if (format == 12)
+      {
+        has_format12 = true;
+        if ((record.platformID == 0 && record.encodingID == 4) ||
+            (record.platformID == 3 && record.encodingID == 10))
+        {
+          unsigned language = (this+record.subtable).u.format12.get_language ();
+          format12_table_languages.set (index, language);
+        }
+      }
+
+      if (record.platformID == 0 && record.encodingID ==  3)
+      {
+        unicode_bmp = table;
+        if (format == 4)
+        {
+          unsigned language = (this+record.subtable).u.format4.get_language ();
+          unicode_bmp_tables.set (language, table);
+        }
+      }
+      else if (record.platformID == 0 && record.encodingID ==  4) unicode_ucs4 = table;
+      else if (record.platformID == 3 && record.encodingID ==  1)
+      {
+        ms_bmp = table;
+        if (format == 4)
+        {
+          unsigned language = (this+record.subtable).u.format4.get_language ();
+          ms_bmp_tables.set (language, table);
+        }
+      }
+      else if (record.platformID == 3 && record.encodingID == 10) ms_ucs4 = table;
+    }
+
+    if (unlikely (!has_format12 && !unicode_bmp && !ms_bmp)) return false;
+    if (unlikely (has_format12 && (!unicode_ucs4 && !ms_ucs4))) return false;
+
+    return true;
+  }
+
+  template<typename Iterator,
+           hb_requires (hb_is_iterator (Iterator))>
+  void prune_redundant_format12_tables (hb_subset_plan_t *plan,
+                                        hb_map_t& format12_table_languages,
+                                        Iterator it,
```

Comment:
Suggest renaming this parameter to unicode_gid _pairs (or something that gives an indication what it iterates over)